components/datadog/agent/docker.go

```diff
@@ -2,31 +2,122 @@ package agent
+	"github.com/DataDog/test-infra-definitions/components/vm"
+	"github.com/DataDog/test-infra-definitions/scenarios/aws/vm/ec2vm"
 )
 
 const (
-	AgentComposeDefinition = `version: "3.9"
-services:
-  agent:
-    image: %s
-    container_name: %s
-    volumes:
-      - "/var/run/docker.sock:/var/run/docker.sock"
-      - "/proc/:/host/proc"
-      - "/sys/fs/cgroup/:/host/sys/fs/cgroup"
-    environment:
-      DD_API_KEY: %s
-      DD_PROCESS_AGENT_ENABLED: true
-      DD_DOGSTATSD_NON_LOCAL_TRAFFIC: true`
 	DefaultAgentImageRepo        = "gcr.io/datadoghq/agent"
 	DefaultClusterAgentImageRepo = "gcr.io/datadoghq/cluster-agent"
 	defaultAgentImageTag         = "latest"
+	AgentContainerName           = "datadog-agent"
 )
 
-func DockerAgentFullImagePath(e *config.CommonEnvironment, repositoryPath string) string {
+type Daemon struct {
+	vm                 *ec2vm.EC2UnixVM
+	dependsOn          pulumi.ResourceOption
+	agentContainerName string
+}
+
+// NewDaemon installs the Datadog Agent through docker-compose on a VM with docker already istalled.
+// It allows running multiple docker applications on the same VM. It does not check if docker is installed,
+// use it with ec2os.AmazonLinuxDockerOS or install docker with `dockerManager.Install` before calling this function.
+func NewDaemon(vm *ec2vm.EC2UnixVM, options ...dockeragentparams.Option) (*Daemon, error) {
+	env := vm.GetCommonEnvironment()
+	params, err := dockeragentparams.NewParams(options...)
+	if err != nil {
+		return nil, err
+	}
+
+	agentFullImagePath := DockerAgentFullImagePath(env, params.Repository, params.ImageTag)
+
+	var composeContents []command.DockerComposeInlineManifest
+	composeContents = append(composeContents, command.DockerComposeInlineManifest{
+		Name:    "agent",
+		Content: dockerAgentComposeContent(agentFullImagePath, env.AgentAPIKey(), params.AgentServiceEnvironment),
+	})
+	composeContents = append(composeContents, params.ExtraComposeManifests...)
+
+	var dependOnResource pulumi.Resource
+	dockerManager := vm.GetLazyDocker()
+
+	if len(composeContents) > 0 {
+		dependOnResource, err = dockerManager.ComposeStrUp("docker-on-vm", composeContents, params.EnvironmentVariables)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &Daemon{
+		vm:                 vm,
+		dependsOn:          utils.PulumiDependsOn(dependOnResource),
+		agentContainerName: AgentContainerName,
+	}, nil
+}
+
+func (d *Daemon) GetDependsOn() pulumi.ResourceOption {
+	return d.dependsOn
+}
+
+func (d *Daemon) GetAgentContainerName() string {
+	return d.agentContainerName
+}
+
+func (d *Daemon) Deserialize(result auto.UpResult) (*ClientData, error) {
+	vmData, err := d.vm.Deserialize(result)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ClientData{Connection: vmData.Connection}, nil
+}
+
+func (d *Daemon) VM() vm.VM {
+	return d.vm.VM
+}
+
+func dockerAgentComposeContent(agentImagePath string, apiKey pulumi.StringInput, envVars pulumi.Map) pulumi.StringOutput {
```

Comment:
I am not sure of the added value of the structured type as it's mixed with untyped content, but LGTM.

Reply:
What do you mean by mixed up with untyped content ?